twilight-gateway/src/ratelimiter.rs

```diff
@@ -1,127 +1,240 @@
+/// Duration until the ratelimit bucket resets.
+const RESET_DURATION: Duration = Duration::from_secs(60);
 
 /// Ratelimiter for sending commands over the gateway to Discord.
 #[derive(Debug)]
 pub struct CommandRatelimiter {
-    /// Bucket used for limiting actions.
-    bucket: LeakyBucket,
+    /// Future that completes the next time the `CommandRatelimiter` allows a
+    /// permit.
+    delay: Pin<Box<Sleep>>,
+    /// Queue of instants started when a command was sent.
+    instants: Vec<Instant>,
 }
 
 impl CommandRatelimiter {
-    /// Create a new ratelimiter.
-    pub(crate) fn new(heartbeat_interval: u64) -> Self {
-        /// Interval of how often to refill the bucket.
-        const REFILL_INTERVAL: Duration = Duration::from_millis(RESET_DURATION_MILLISECONDS);
-
-        // Number of commands allotted to the user per reset period.
-        let commands_allotted = u32::from(available_commands_per_interval(heartbeat_interval));
-
-        let bucket = LeakyBucket::builder()
-            .max(commands_allotted)
-            .tokens(commands_allotted)
-            .refill_interval(REFILL_INTERVAL)
-            .refill_amount(commands_allotted)
-            .build();
-
-        Self { bucket }
+    /// Create a new ratelimiter with some capacity reserved for heartbeating.
+    pub(crate) fn new(heartbeat_interval: Duration) -> Self {
+        let allotted = nonreserved_commands_per_reset(heartbeat_interval);
+
+        Self {
+            delay: Box::pin(sleep(Duration::ZERO)),
+            instants: Vec::with_capacity(allotted.into()),
+        }
     }
 
     /// Current number of commands that are still available within the interval.
-    pub fn available(&self) -> u32 {
-        self.bucket.tokens()
+    pub fn available(&self) -> u8 {
+        // filter out elapsed instants
+        let used_permits: u8 = self
+            .instants
+            .iter()
+            .filter(|instant| instant.elapsed() < RESET_DURATION)
+            .count()
+            .try_into()
+            .expect("length is at most 118");
+
+        self.max() - used_permits
     }
 
     /// Maximum number of commands that may be made per interval.
-    pub fn max(&self) -> u32 {
-        self.bucket.max()
+    #[allow(clippy::cast_possible_truncation)]
+    pub fn max(&self) -> u8 {
+        self.instants.capacity() as u8
     }
 
-    /// When the bucket will refresh the available number of commands again.
-    pub fn next_refill(&self) -> Instant {
-        self.bucket.next_refill().into_std()
+    /// When the next command is available.
+    pub fn next_available(&self) -> Duration {
+        self.instants.first().map_or(Duration::ZERO, |instant| {
+            RESET_DURATION.saturating_sub(instant.elapsed())
+        })
     }
 
-    /// Acquire a token from the bucket, waiting until one is available.
-    pub(crate) async fn acquire_one(&self) {
-        self.bucket.acquire_one().await;
+    /// Completes when a ratelimit permit is available.
```

Comment:
i think this doc could be better:
```suggestion
    /// Returns a ratelimit permit when one becomes available.
```